### PR TITLE
[Networking] Refactoring networking received cache to HeroCache

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This PR implements an alternative `RcvCache` cache based on `HeroCache`. It then removes the existing `hashicorp`-based networking received cache and reactors the code to utilize the `HeroCache`-based one.  

Note: `HeroCache` is our homebrewed cache which is around 4 times faster than `stdmap` version and incures `O(1)` heap objects.

### CODE REVIEW

File: network/p2p/network.go [hunk 1]
```diff
@@ -333,7 +330,7 @@ func (n *Network) Receive(nodeID flow.Identifier, msg *message.Message) error {
 
 func (n *Network) processNetworkMessage(senderID flow.Identifier, message *message.Message) error {
 	// checks the cache for deduplication and adds the message if not already present
-	if n.rcache.add(message.EventID, network.Channel(message.ChannelID)) {
```
Comment (yhassanzadeh13) [inline network/p2p/network.go:336]: `EventID` also encapsulates the channel.

File: network/p2p/network.go [hunk 2]
```diff
@@ -333,7 +330,7 @@ func (n *Network) Receive(nodeID flow.Identifier, msg *message.Message) error {
...
+	if !n.rcache.Add(message.EventID) {
```
Comment (yhassanzadeh13) [inline network/p2p/network.go:333]: To be consistent with other `Add` interfaces in the codebase, the `Add` returns true for unseen messages.

File: network/cache/rcvcache.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	herocache "github.com/onflow/flow-go/module/mempool/herocache/backdata"
+	"github.com/onflow/flow-go/module/mempool/herocache/backdata/heropool"
+	"github.com/onflow/flow-go/module/mempool/stdmap"
+)
+
+// RcvCache implements an LRU cache of the received eventIDs that delivered to their engines
+type RcvCache struct {
```
Comment (vishalchangrani) [inline network/cache/rcvcache.go:13]: nit: maybe rename it to `ReceiveCache` to be easy for anyone new to the codebase

File: network/p2p/network.go [hunk 4]
```diff
@@ -545,3 +533,19 @@ func (n *Network) queueSubmitFunc(message interface{}) {
 
 	n.metrics.InboundProcessDuration(qm.Target.String(), time.Since(startTimestamp))
 }
+
+func EventId(channel network.Channel, payload []byte) (hash.Hash, error) {
+	// use a hash with an engine-specific salt to get the payload hash
+	h := hash.NewSHA3_384()
+	_, err := h.Write([]byte("libp2ppacking" + channel))
```
Comment (vishalchangrani) [inline network/p2p/network.go:540]:
```suggestion
        const prefix = "libp2ppacking"
	_, err := h.Write([]byte(preifx + channel))
```

Comment (vishalchangrani) [reply network/p2p/network.go:540]: maybe golang is a lil more efficient with constants than a string being define on each method call

File: network/cache/rcvcache.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package netcache
```
Comment (vishalchangrani) [inline network/cache/rcvcache.go:1]: thanks for putting all this in a cache package.

**Review by vishalchangrani — APPROVED** [review]

File: network/cache/rcvcache.go (hunk 3 above)
Comment (yhassanzadeh13) [reply network/cache/rcvcache.go]: https://github.com/onflow/flow-go/pull/1908/commits/acb7563b456c7e30fbb02ac552bf6565d2b04ad7

File: network/p2p/network.go (hunk 4 above)
Comment (yhassanzadeh13) [reply network/p2p/network.go]: https://github.com/onflow/flow-go/pull/1908/commits/603f57f7ae1f2bc17bc7fc61d55375d501b6dc32

**Review by simonhf — APPROVED** [review]
LGTM :-)

**Pull request merged**